### PR TITLE
[WIP] SlimCUDSSource for low-memory VTK CUDS 

### PR DESCRIPTION
**Pull request by stefanoborini**
Extracted what I learned into a "best compromise" solution. Still missing tests and a serious check.

### CODE REVIEW

**Comment by kitchoi** [comment]
Is `CubaData.insert` useful here?

File: simphony_mayavi/sources/slim_cuds_source.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        # VTKDataSource _from_ the VTK data, but our VTK data will not contain
+        # all the keys. We are therefore required to update them here.
+        all_lists = (
+            self._point_scalars_list,
+            self._point_vectors_list,
+            self._cell_scalars_list,
+            self._cell_vectors_list)
+
+        for lst, keys in zip(all_lists, _available_keys(cuds)):
+            lst[:] = sorted([key.name for key in keys])
```
Comment (kitchoi) [inline simphony_mayavi/sources/slim_cuds_source.py]: Need to add an empty string here for the user to unselect the data attribute:
https://github.com/enthought/mayavi/blob/master/mayavi/sources/vtk_data_source.py#L345

File: simphony_mayavi/sources/slim_cuds_source.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+from simphony.io.h5_mesh import H5Mesh
+from simphony.testing.utils import dummy_cuba_value
+from simphony_mayavi.cuds.vtk_lattice import VTKLattice
+from simphony_mayavi.cuds.vtk_mesh import VTKMesh
+from simphony_mayavi.cuds.vtk_particles import VTKParticles
+from simphony_mayavi.sources.cuds_source import CUDSSource
+
+
+class SlimCUDSSource(CUDSSource):
+    """Overrides CUDSSource to provide a source optimized for lower memory
```
Comment (itziakos) [inline simphony_mayavi/sources/slim_cuds_source.py:22]: Can I suggest to create a base `BaseCUDSSource` and having `CUDSSource`, `SlimCUDSSource` subclass from that?

File: simphony_mayavi/sources/slim_cuds_source.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+    Returns
+    -------
+    scalars, vectors : tuple(set, set)
+        The CUBA type keys for scalar and vector data, respectively
+    """
+    scalars = set()
+    vectors = set()
+
+    for cuba_key in data.keys():
```
Comment (itziakos) [inline simphony_mayavi/sources/slim_cuds_source.py]: you do not have to iterate over all the keys just the supported cuba, https://github.com/simphony/simphony-mayavi/blob/master/simphony_mayavi/core/cuba_utils.py#L8

File: simphony_mayavi/sources/slim_cuds_source.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        try:
+            dim = _cuba_dimensionality(cuba_key)
```
Comment (itziakos) [inline simphony_mayavi/sources/slim_cuds_source.py]: You can probably get the dimensionality of the CUBA key from `KEYWORDS`, https://github.com/simphony/simphony-common/blob/master/simphony/core/keywords.py

File: simphony_mayavi/sources/slim_cuds_source.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+    else:
+        msg = 'Provided object {} is not of any known cuds type'
+        raise TraitError(msg.format(type(cuds)))
+
+    return point_scalars, point_vectors, cell_scalars, cell_vectors
+
+
+def _extract_cuba_keys_per_data_types(data):
+    """Given a DataContainer, gets the scalar and vector CUBA types
```
Comment (itziakos) [inline simphony_mayavi/sources/slim_cuds_source.py:298]: suggestion: You should be able to precalculate a mapping between supported keys in simphony_mayavi and vtk data type correspondance. The you just get the intersection with the data.keys()

File: simphony_mayavi/sources/slim_cuds_source.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    """Small generator that processes an iterable of strings
+    into an iterable of CUBA keys"""
+    for elem in iterable:
+        try:
+            yield CUBA[elem]
+        except ValueError:
+            pass
+
+
+def _available_keys(cuds):
```
Comment (itziakos) [inline simphony_mayavi/sources/slim_cuds_source.py:230]: utility function do not have to be private. However about moving them to the core module

File: simphony_mayavi/sources/slim_cuds_source.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+    def _update_data(self):
+        # We need to silence the behavior of the VTKDataSource, otherwise it
+        # will overwrite the point_scalar_list etc. with data from the vtk
+        # datasource, which in this case contains only the data we care
+        # about.
+        pass
+
+
+def _to_cuba(iterable):
```
Comment (itziakos) [inline simphony_mayavi/sources/slim_cuds_source.py]: This is not used

File: simphony_mayavi/sources/slim_cuds_source.py (hunk 3 above)
Comment (stefanoborini) [inline simphony_mayavi/sources/slim_cuds_source.py]: I don't want to know the dimensionality of generic cuba data. I want to know the dimensionality of the data that is actually stored in that specific datacontainer.

Comment (itziakos) [inline simphony_mayavi/sources/slim_cuds_source.py]: Sorry I wanted to say iterate, over the intersection data.keys() and supported_cuba()

File: simphony_mayavi/sources/slim_cuds_source.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    -------
+    scalars, vectors : tuple(set, set)
+        The CUBA type keys for scalar and vector data, respectively
+    """
+    scalars = set()
+    vectors = set()
+
+    for cuba_key in data_container.keys():
+        try:
+            shape = _cuba_shape(cuba_key)
```
Comment (itziakos) [inline simphony_mayavi/sources/slim_cuds_source.py]: If the keys are from a data_container how is it possible that `_cuba_shape` raises a ValueError? All keys in a data container are CUBA keys

File: simphony_mayavi/sources/slim_cuds_source.py (hunk 2 above)
Comment (stefanoborini) [inline simphony_mayavi/sources/slim_cuds_source.py:22]: After it's tested I'll do this refactoring

**Comment by stefanoborini** [comment]
This implementation seems to work, but its code is awful. I just added a few more tests to spot issues and now I create a branch out of this last commit to perform a refactoring, now that we understand a bit more the logic involved. If I run out of time, at least we have this PR available for merging.

**Comment by itziakos** [comment]
> If I run out of time, at least we have this PR available for merging.

If you are not happy with the code the please do not merge. It would be better to cancel this PR and make the new branch. Alternatively you can do the refactoring/cleanup here.

File: simphony_mayavi/sources/slim_cuds_source.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+            point_vectors.update(vector_keys)
+
+    else:
+        msg = 'Provided object {} is not of any known cuds type'
+        raise TraitError(msg.format(type(cuds)))
+
+    return point_scalars, point_vectors, cell_scalars, cell_vectors
+
+
+def _extract_cuba_keys_per_data_types(data_container):
```
Comment (itziakos) [inline simphony_mayavi/sources/slim_cuds_source.py:297]: Please consider https://github.com/simphony/simphony-mayavi/pull/171#discussion_r55885605

**Comment by stefanoborini** [comment]
Closing. Needs refactoring.

**Pull request closed without merging**